File: julia/algorithms/mergesort.jl
function mergesort(array_in)

    n = length(array_in)

    if n <= 1
        return array_in
    else
        subarray1 = array_in[1:ceil(n/2)]
        subarray2 = array_in[ceil(n/2)+1:n]
        a = mergesort(subarray1)
        b = mergesort(subarray2)
        array_out = qmerge(a, b)
        return array_out
    end

end

function qmerge(array1, array2)

    imax = length(array1)
    jmax = length(array2)
    kmax = imax + jmax + 1

    array_out = zeros(imax+jmax)

    i = 1
    j = 1
    k = 1

    while k < kmax

        # Check if at end of either array
        if i == (imax + 1)
            array_out[k:end] = array2[j:end]
            return array_out
        end

        if j == (jmax + 1)
            array_out[k:end] = array1[i:end]
            return array_out
        end

        # Standard Case
        if array1[i] < array2[j]
            array_out[k] = array1[i]
            i += 1
            k += 1
        else
            array_out[k] = array2[j]
            j += 1
            k += 1
        end
    end
end

File: scripting/python/algorithms/MergeSort.py
# -*- coding: utf-8 -*-
"""
Created on Thu May  1 21:03:14 2014

@author: tdoughty1
"""

from numpy import zeros, random
from time import time


def QMerge(array1, array2):

    i = 0
    j = 0
    k = 0

    imax = len(array1)
    jmax = len(array2)
    kmax = imax + jmax

    tempArray = zeros(kmax)

    while k < kmax:

        # At end of first array
        if i == imax:
            tempArray[k:] = array2[j:]
            return tempArray

        # At end of second array
        if j == jmax:
            tempArray[k:] = array1[i:]
            return tempArray

        # Standard Case
        if array1[i] < array2[j]:
            tempArray[k] = array1[i]
            i += 1
            k += 1
        else:
            tempArray[k] = array2[j]
            j += 1
            k += 1


def MergeSort(testArray):

    n = len(testArray)

    if n == 1:
        return testArray

    else:
        array1 = testArray[0:n/2]
        array2 = testArray[n/2:]

        array1 = MergeSort(array1)
        array2 = MergeSort(array2)

        return QMerge(array1, array2)

N = 1000000

testArray = random.rand(N)

start = time()
testArray = MergeSort(testArray)
finish = time()

print "Took %.3f us" % ((finish-start)*1000)
